library/src/armory/export/criteria.py

```diff
@@ -0,0 +1,512 @@
+        return [sample_idx for sample_idx in range(len(batch)) if sample_idx < n]
+
+    return _criterion
+
+
+def first_n_samples(n: int) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches the first n samples in the dataset.
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import first_n_samples
+
+        # Exports the first 5 samples, regardless of batch
+        # For example with a batch size of 2:
+        #  - batch 0, samples 0 and 1
+        #  - batch 1, samples 0 and 1
+        #  - batch 2, sample 0
+        exporter = Exporter(criterion=first_n_samples(5))
+
+    Args:
+        n: Number of samples to be exported
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if n == 0:
+            return False
+        batch_size = len(batch)
+        if batch_idx * batch_size >= n:
+            return False
+        return [
+            sample_idx
+            for sample_idx in range(batch_size)
+            if ((batch_idx * batch_size) + sample_idx) < n
+        ]
+
+    return _criterion
+
+
+def chains(names: Sequence[str]) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches all samples in every batch for a
+    particular set of perturbation chains.
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import chains
+
+        # Exports every sample in every batch for chains "benign" and "defended"
+        exporter = Exporter(criterion=chains(["benign", "defended"]))
+
+    Args:
+        names: Names of perturbation chains from which to export
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name: str, batch_idx, batch):
+        return chain_name in names
+
+    return _criterion
+
+
+def samples(indices: Sequence[int]) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches specific samples in the dataset, by
+    their global index (regardless of batch).
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import samples
+
+        # For a batch size of 4:
+        #  - batch 0, sample 2
+        #  - batch 1, samples 0 and 1
+        #  - batch 2, samples 2 and 3
+        exporter = Exporter(criterion=samples([2, 4, 5, 10, 11]))
+
+    Args:
+        indices: Indices of samples within the dataset to be exported
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if len(indices) == 0:
+            return False
+        batch_size = len(batch)
+        return [
+            sample_idx
+            for sample_idx in range(batch_size)
+            if ((batch_idx * batch_size) + sample_idx) in indices
+        ]
+
+    return _criterion
+
+
+def _create_metric_criterion(comp, metric, threshold) -> Exporter.Criterion:
```

Comment:
With type hints and some documentation, this function would be useful as a user utility to create custom metric criteria.

Reply:
Since the user is already in complete control of the `metric` itself, making this public would only add the ability for the user to specify custom `comp` functions--outside of standard/typical comparisons (e.g., eq/ne/lt/lte/gt/gte), are there actually any additional comparisons the user would find useful?

Reply:
Right, in that example, the user would just need to have their `predicate` function call `nonzero().flatten().tolist()` on the tensor before returning. I'm inclined to omit such a utility unless we find it to be a common occurrence (and at that point, the better alternative may be to allow all criterion functions to return `Union[bool, Iterable[int], torch.Tensor]` and do the nonzero flatten in the base exporter).